common-rest/src/main/java/org/uniprot/api/rest/controller/BasicSearchController.java

```diff
@@ -114,26 +115,35 @@ protected BasicSearchController(
         } else {
             context.setEntities(result);
         }
-
-        return getResponseBodyEmitterResponseEntity(request, context);
+        return getDeferredResultResponseEntity(request, context);
     }
 
-    protected ResponseEntity<ResponseBodyEmitter> getResponseBodyEmitterResponseEntity(
-            HttpServletRequest request, MessageConverterContext<T> context) {
-        ResponseBodyEmitter emitter = new ResponseBodyEmitter();
+    protected DeferredResult<ResponseEntity<MessageConverterContext<T>>>
+            getDeferredResultResponseEntity(
+                    HttpServletRequest request, MessageConverterContext<T> context) {
+
+        // timeout in millis
+        Long timeoutInMillis = Long.valueOf(downloadTaskExecutor.getKeepAliveSeconds()) * 1000;
+
+        DeferredResult<ResponseEntity<MessageConverterContext<T>>> deferredResult =
+                new DeferredResult<>(timeoutInMillis);
+
+        deferredResult.onTimeout(() -> log.error("Request timeout occurred."));
+
         downloadTaskExecutor.execute(
                 () -> {
                     try {
-                        emitter.send(context, context.getContentType());
-                    } catch (IOException e) {
-                        emitter.completeWithError(e);
+                        deferredResult.setResult(
```

Comment:
Just thinking, do you test what happens when a client does CTRL-C, to abort a download? It'd be good to test to make sure the deferred result handles it gracefully and doesn't wait / consume resources.

Reply:
It closes gracefully. The subsequent requests work fine. I need to check if it still keeps some resources open